pandas/core/apply.py

```diff
@@ -340,7 +340,10 @@ def wrap_results_for_axis(
 
         if self.result_type == "reduce":
             # e.g. test_apply_dict GH#8735
-            return self.obj._constructor_sliced(results)
+            res = self.obj._constructor_sliced(results)
+            res.index = res_index
```

Comment:
Any reason you don't just pass the index to the constructor?

Reply:
because we have e.g. `results = {0: "foo", 1: "bar"}` and `res_index = Index(["A", "B"])`, so passing index to the constructor leads to all-NAss